### PR TITLE
OpenEnclave PAL: Store enclave heap base/end in inline variables.

### PR DESCRIPTION
**Pull request by anakrish**
PALOpenEnclave object is lazily constructed. I couldn't
figure out a straight-forward way to pass the heap bounds to
the constructor of PALOpenEnclave object.
As an alternative, store the bounds in class-static members of a template
class.

- two_alloc_types/main.cc
  OE use of snmalloc is renamed to snmalloc_enclave.
  Declare oe_base and oe_end inline variables in snmalloc_enclave
  namespace and initialize them.

- fixed_region/fixed_region.cc
  Initialize oe_base and oe_end variables.
  Cache the starting value of oe_base in orig_oe_base and
  use that for asserting that the allocated memory lies
  within the PAL heap. This is because, in pal_open_enclave.h,
  reserve will change the value of oe_base.

CC: @achamayou 
Signed-off-by: Anand Krishnamoorthi <anakrish@microsoft.com>

### CODE REVIEW

File: src/pal/pal_open_enclave.h [hunk 1]
```diff
@@ -9,6 +9,20 @@ extern "C" [[noreturn]] void oe_abort();
 
 namespace snmalloc
 {
+  // Use a template to hold the heap base and end variables.
+  // Templates prevent multiple-definition error for these variables,
+  // without having to use a .cpp file otherwise.
+  template<typename T>
+  struct HeapBounds
+  {
+    static std::atomic<void*> heap_base;
```
Comment (mjp41) [inline src/pal/pal_open_enclave.h:18]: We can use C++ inline statics here to remove the need for the template.
```C++
inline static std::atomic<void*> heap_base;
```

File: src/pal/pal_open_enclave.h [hunk 2]
```diff
@@ -9,6 +9,20 @@ extern "C" [[noreturn]] void oe_abort();
...
+    static void* heap_end;
+  };
+  template<typename T>
+  std::atomic<void*> HeapBounds<T>::heap_base;
+  template<typename T>
+  void* HeapBounds<T>::heap_end;
+
   class PALOpenEnclave
   {
     std::atomic<void*> oe_base = nullptr;
```
Comment (mjp41) [inline src/pal/pal_open_enclave.h:28]: This fields should be dropped.

Comment (anakrish) [reply src/pal/pal_open_enclave.h]: Fixed.

File: src/pal/pal_open_enclave.h (hunk 1 above)
Comment (anakrish) [reply src/pal/pal_open_enclave.h]: I had to use `inline` variables instead of `inline static` so that I could initialize these in src/test/func/two_alloc_types/main.cc

**Comment by anakrish** [comment]
@mjp41 Thanks for the suggestions. Can you review the latest iteration?

File: src/pal/pal_open_enclave.h (hunk 1 above)
Comment (mjp41) [reply src/pal/pal_open_enclave.h]: I think `inline static` would be required if you kept the struct. But the `inline` global is perfect here.  Thanks for addressing.

File: src/pal/pal_open_enclave.h [hunk 3]
```diff
@@ -2,17 +2,18 @@
...
 #include "pal_plain.h"
 #ifdef OPEN_ENCLAVE
-extern "C" const void* __oe_get_heap_base();
-extern "C" const void* __oe_get_heap_end();
 extern "C" void* oe_memset_s(void* p, size_t p_size, int c, size_t size);
 extern "C" [[noreturn]] void oe_abort();
 
 namespace snmalloc
 {
+  // The following bounds variables will be setup via the
```
[the remark this answers is not in the document]
Comment (anakrish) [reply src/pal/pal_open_enclave.h]: Nice suggestion. I have incorporated it.

File: src/test/func/two_alloc_types/main.cc [hunk 4]
```diff
@@ -1,22 +1,11 @@
 #include "../../../snmalloc.h"
 
+#include <atomic>
```
Comment (mjp41) [inline src/test/func/two_alloc_types/main.cc:3]: Is this change required?

Comment (anakrish) [reply src/test/func/two_alloc_types/main.cc]: Good catch. Removed.

**Comment by anakrish** [comment]
> Thanks for updating it. Just a minor clangformat issue, then this is good to go.
> 
> You can do
> 
> ```
> ninja clangformat
> ```
> 
> to run clangformat with the relevant settings.

Done. Looks like the CI is happy too.
